devito/types/equation.py

```diff
@@ -156,6 +161,58 @@ def __str__(self):
     __repr__ = __str__
 
 
+class Break(Eq):
```

Comment:
What is the purpose of this class? Can that be avoided?

Reply:
I also discussed the purpose of the Break equation with Nelson. It will be used to allow halting the execution of an operator while in the middle of its execution. Currently all the loop iterations must be executed before exiting the Operator call. Let's say we want to stop the operator execution if we find out that the energy is getting dissipated or after a residual criteria has been met. Currently, the only way to do that is to execute some iterations, go back to python, check for the condition and then come back to C code. The Break equation introduces a way to implement iterative solvers since it allows checking for a condition and breaking the loop inside C code.

Reply:
btw, should this really be called `Break`? 

why can't we just have `condition` has an additional kwarg of `Eq`, thus avoiding the significant amount of addition that this PR is trying to bring in?

Reply:
The initial reason behind making a new class is that `Break` is not actually an equation, but an inequation. One thing to notice too is that a Break condition is not a property of a single equation, but is a property of a set of equation that should stop executing whenever a condition is met.

So if implementing using this `condition` property, the actual Break instruction would look something like `brkEq = Eq(condition=residual<epsilon)`, is that OK for you? I will try to rewrite everything using that idea, which should avoid a lot of code indeed.

Reply:
I think we have a number of possibilities and we should explore them all. I'm not sure which one is best yet, and I'd like feedback from the more seasoned developers (@rhodrin @mloubout )

> but is a property of a set of equation that

fair point. 

```
brkEq = Eq(break=residual<epsilon)
```

yes, I think this could work.  Thoughts?

I'm still a bit reluctant about the chosen name. How about `condition` ? 

I also have another question. How is this going to interact with [ConditionalDimension](http://devitocodes.github.io/devito/dimension.html#devito.types.dimension.ConditionalDimension) ? 

* Is this a completely separate abstraction? 
* Both will generate an `if-then` eventually, so can we, and if so, how are we going to, share the common bits of the compilation pipeline? For example, ConditionalDimensions are attached [here](https://github.com/devitocodes/devito/blob/master/devito/ir/equations/equation.py#L159) and later processed and attached to Clusters [here](https://github.com/devitocodes/devito/blob/master/devito/ir/clusters/algorithms.py#L287)

Again, I'm looking for feedback here

Reply:
I should have thought of and mentioned this before, but Sympy has inequality solvers right? Shouldn't we be 'mimicking' that api if possible? I'll look into them as post feedback.